### PR TITLE
[Spells] New SPA focus effects and focus limits

### PR DESCRIPTION
**Pull request by KayenEQ**
Implemented various new focus effects and focus limiters (7/20/21)

- Added basic standardized focus and limit descriptions to source in spdat.h, on other areas of relevance. Should improve
understandability of the effects. @Fc = Focus @Ff = Focus Limiters
- Minor fixes to  a couple of AA spell limits
---------------------------------------------------------------------->
Foucus Effects

SE_Fc_Amplify_Mod 507  On Caster, damage-heal-dot mod pct, base: pct

SE_Fc_Amplify_Amt 508 On Caster, damage-heal-dot mod flat amt, base: amt

SE_Fc_ResistIncoming 510 On Target, resist modifier, base: amt 
Note: 2 pts = 1% resist mod +/-

SE_Fc_CastTimeMod2 500  On Caster, cast time mod pct, base: pct
Note: There is no restriction to much this can lower spell time, if lowered beyond 0, then will instant cast.

SE_Fc_CastTimeAmt  501  On Caster, cast time mod flat amt, base: milliseconds. 
Note: There is no restriction to much this can lower spell time, if lowered beyond 0, then will instant cast.

---------------------------------------------------------------------->
Foucs Limiters

SE_Ff_DurationMax	 495 Max duration of spell that can be focused, base: tics

SE_Ff_ReuseTimeMin  490 Minimum recast time of a spell that can be focused, base: recast time

SE_Ff_ReuseTimeMax  491 Max recast time of a spell that can be focused, base: recast time

SE_Ff_Endurance_Min 492 Minimum endurance cost of a spell that can be focused, base: endurance cost

SE_Ff_Endurance_Max 493 Max endurance cost of a spell that can be focused, base: endurance cost

SE_Ff_Value_Min 479 Minimum base value of a spell that can be focused, base: spells to be focused base1 value

SE_Ff_Value_Max 480 Max base value of a spell that can be focused, base: spells to be focused base1 value

SE_Ff_Override_NotFocusable	 460 Allow spell to be focused event if flagged with 'not_focusable' in spell table, base: 1

### CODE REVIEW

**Comment by splose** [comment]
Will test these tomorrow.

492/493 and 500/501 will be very helpful for the Synergy AA's I ported over. Some of the melee's are held together with sticks and bubblegum at the moment. :heart:

**Review by splose — APPROVED** [review]

**Comment by mackal** [comment]
This looks fine, but I'm a bit concerned with stacking order of the new SPA 500 and 501, if anyone can give definitive proof that would be good. (looks like easiest test would be level 110 bard AA Quick Time, just blow the AA in a buff suspend zone and check how the cast time changes removing gear etc)

**Comment by Akkadius** [comment]
This looks dope!

Per our DM conversation; we'll jump on a call to discuss some of the items you were asking about

File: zone/mob.cpp [hunk 1]
```diff
@@ -3120,16 +3120,22 @@ uint32 Mob::GetLevelHP(uint8 tlevel)
...
 	    (GetClass() == RANGER || GetClass() == SHADOWKNIGHT || GetClass() == PALADIN || GetClass() == BEASTLORD)) {
 		int level_mod = std::min(15, GetLevel() - 50);
 		cast_reducer += level_mod * 3;
 	}
 
+	cast_reducer = std::min(cast_reducer, 50);  //Max cast time with focusSpellHaste and level reducer is 50% of cast time.
+	cast_reducer += cast_reducer_no_limit;
 	casttime = casttime * (100 - cast_reducer) / 100;
-	return std::max(casttime, min_cap);
+	casttime -= cast_reducer_amt;
```
Comment (mackal) [inline zone/mob.cpp:3135]: Until I can get around to proving this happens at a different time, I think I'm gonna say this is fine.

File: zone/spell_effects.cpp [hunk 2]
```diff
@@ -5181,6 +5274,37 @@ int16 Mob::CalcFocusEffect(focusType type, uint16 focus_id, uint16 spell_id, boo
...
 
+		case SE_Ff_DurationMax:
+			if (focus_spell.base[i] > spell.buffduration)
+				return 0;
+			break;
+
+		case SE_Ff_Endurance_Min:
+			if (spell.EndurCost < focus_spell.base[i])
+				return 0;
+
```
Comment (Akkadius) [inline zone/spell_effects.cpp:5285]: Missing breaks in a lot of these cases

**Review by Akkadius — CHANGES_REQUESTED** [review]
Minor changes, LGTM

File: zone/spell_effects.cpp (hunk 2 above)
Comment (KayenEQ) [reply zone/spell_effects.cpp:5286]: resolved.

**Review by Akkadius — APPROVED** [review]

**Pull request merged**